### PR TITLE
PWGHF: Developments for Charm Resonances Reduced workflow

### PR DESCRIPTION
**Pull request by Luca610**
ReducedDataModel.h: 
- Added event selection columns to CollisionCounter table
- Added Daughters momenta to V0 and 3prongNoTrk tables
- Changed invariant mass and momentum columns to dynamic

Changes in DataCreators and CandidateCreators to work with new tables

### CODE REVIEW

File: PWGHF/D2H/DataModel/ReducedDataModel.h [hunk 1]
```diff
@@ -435,33 +435,108 @@ DECLARE_SOA_COLUMN(MlScoreBkgProng0, mlScoreBkgProng0, float);             //! B
 DECLARE_SOA_COLUMN(MlScorePromptProng0, mlScorePromptProng0, float);       //! Prompt ML score of the D daughter
 DECLARE_SOA_COLUMN(MlScoreNonpromptProng0, mlScoreNonpromptProng0, float); //! Nonprompt ML score of the D daughter
 } // namespace hf_reso_cand_reduced
+namespace hf_reso_3_prong
+{
+DECLARE_SOA_COLUMN(PxProng0, pxProng0, float);                              
+DECLARE_SOA_COLUMN(PyProng0, pyProng0, float);
+DECLARE_SOA_COLUMN(PzProng0, pzProng0, float);
+DECLARE_SOA_COLUMN(PxProng1, pxProng1, float);
```
Comment (vkucera) [inline PWGHF/D2H/DataModel/ReducedDataModel.h]: Most of your columns already exist in `CandidateReconstructionTables`. Why do you need to duplicate them?

File: PWGHF/D2H/TableProducer/candidateCreatorCharmResoReduced.cxx [hunk 2]
```diff
@@ -212,7 +219,7 @@ struct HfCandidateCreatorCharmResoReduced {
         }
         float invMassReso{0.};
         float invMassV0{0.};
-        std::array<float, 3> pVecV0 = candV0.pVector();
+        std::array<float, 3> pVecV0 = {candV0.px(), candV0.py(), candV0.pz()};
```
Comment (vkucera) [inline PWGHF/D2H/TableProducer/candidateCreatorCharmResoReduced.cxx:227]: Why are you reverting this?

File: PWGHF/D2H/TableProducer/dataCreatorCharmHadPiReduced.cxx [hunk 3]
```diff
@@ -824,19 +826,28 @@ struct HfDataCreatorCharmHadPiReduced {
...
+      if (!TESTBIT(rejectionMask, o2::hf_evsel::EventRejection::PositionZ)) zvtxColl++;
+      if (!TESTBIT(rejectionMask, o2::hf_evsel::EventRejection::PositionZ) && !TESTBIT(rejectionMask, o2::hf_evsel::EventRejection::Trigger) ) zvtxAndSel8Coll++;
+      if (rejectionMask == 0) allSelColl++;
       auto thisCollId = collision.globalIndex();
       auto candsCThisColl = candsC.sliceBy(candsDplusPerCollision, thisCollId);
       auto trackIdsThisCollision = trackIndices.sliceBy(trackIndicesPerCollision, thisCollId);
       runDataCreation<false, false, DecayChannel::B0ToDminusPi>(collision, candsCThisColl, trackIdsThisCollision, tracks, tracks, bcs);
     }
+    // handle normalization by the right number of collisions
+    hfCollisionCounter(collisions.tableSize(), zvtxColl, sel8Coll, zvtxAndSel8Coll, allSelColl );
```
Comment (vkucera) [inline PWGHF/D2H/TableProducer/dataCreatorCharmHadPiReduced.cxx]: This block is now duplicated many times. It would be better to factor it out.

**Review by vkucera — CHANGES_REQUESTED** [review]

File: PWGHF/D2H/DataModel/ReducedDataModel.h (hunk 1 above)
Comment (Luca610) [reply PWGHF/D2H/DataModel/ReducedDataModel.h]: No real need, I just wanted to have all the columns in the same place, but I can use the existing ones if needed

File: PWGHF/D2H/TableProducer/candidateCreatorCharmResoReduced.cxx (hunk 2 above)
Comment (Luca610) [reply PWGHF/D2H/TableProducer/candidateCreatorCharmResoReduced.cxx:227]: I deleted the PVector column for that table

File: PWGHF/D2H/TableProducer/dataCreatorCharmHadPiReduced.cxx (hunk 3 above)
Comment (Luca610) [reply PWGHF/D2H/TableProducer/dataCreatorCharmHadPiReduced.cxx]: Working on it! Should I make it into a small function in "utilsEvSelHf.h"?

File: PWGHF/D2H/TableProducer/candidateCreatorCharmResoReduced.cxx (hunk 2 above)
Comment (vkucera) [reply PWGHF/D2H/TableProducer/candidateCreatorCharmResoReduced.cxx:227]: Why?

File: PWGHF/D2H/TableProducer/dataCreatorCharmHadPiReduced.cxx (hunk 3 above)
Comment (vkucera) [reply PWGHF/D2H/TableProducer/dataCreatorCharmHadPiReduced.cxx]: Small function yes, but I would not put it in `utilsEvSelHf.h` as this is specfici to the reduced data model.

File: PWGHF/D2H/TableProducer/candidateCreatorCharmResoReduced.cxx (hunk 2 above)
Comment (Luca610) [reply PWGHF/D2H/TableProducer/candidateCreatorCharmResoReduced.cxx:227]: I used dynamic columns for the candidate's momentum components, so I couldn't use them as argument of the PVector column anymore. I could have defined a new dynamic column that takes the daughter tracks' momenta as an input...

File: PWGHF/D2H/TableProducer/dataCreatorCharmHadPiReduced.cxx (hunk 3 above)
Comment (Luca610) [reply PWGHF/D2H/TableProducer/dataCreatorCharmHadPiReduced.cxx]: Ok

**Comment by Luca610** [comment]
Dear @vkucera  thank you for your comments! In the new version of this PR I replaced duplicate static columns with existing ones from CandidateReconstructionTables.h, and made a function for the checks on event selection

File: PWGHF/D2H/TableProducer/dataCreatorCharmHadPiReduced.cxx [hunk 4]
```diff
@@ -244,6 +246,25 @@ struct HfDataCreatorCharmHadPiReduced {
     setLabelHistoCands(hCandidatesDPlus);
   }
 
+  /// Helper function to count collisions at different event selection stages
+  /// \tparam useEvSel use information from the EvSel table
+  /// \tparam centEstimator centrality estimator
+  /// \param collision collision to test against the selection criteria
+  template <bool useEvSel, o2::hf_centrality::CentralityEstimator centEstimator, typename Coll>
+  void checkEvSel(Coll collision, int& zvtxColl, int& sel8Coll, int& zvtxAndSel8Coll, int& allSelColl)
```
Comment (vkucera) [inline PWGHF/D2H/TableProducer/dataCreatorCharmHadPiReduced.cxx:254]:
```suggestion
  void checkEvSel(Coll const& collision, int& zvtxColl, int& sel8Coll, int& zvtxAndSel8Coll, int& allSelColl)
```

**Review by vkucera — CHANGES_REQUESTED** [review]

**Comment by vkucera** [comment]
> Dear @vkucera thank you for your comments! In the new version of this PR I replaced duplicate static columns with existing ones from CandidateReconstructionTables.h, and made a function for the checks on event selection

Thanks, it looks better now, but you still have 3 identical definitions of the same function. I would indeed suggest to make a header for utilities relevant to the reduced data model.

**Comment by Luca610** [comment]
Dear @vkucera thanks again for your inputs. In this version I moved the event selection function to a dedicated header file to be expanded with future developments for the reduced data format workflows

File: PWGHF/D2H/Utils/utilsRedDataFormat.h [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+#ifndef PWGHF_D2H_UTILS_UTILSREDDATAFORMAT_H_
+#define PWGHF_D2H_UTILS_UTILSREDDATAFORMAT_H_
+
+#include "PWGHF/Core/CentralityEstimation.h"
+#include "PWGHF/Utils/utilsEvSelHf.h"
+
+namespace o2::hf_evsel
+{
+HfEventSelection hfEvSel;
```
Comment (fgrosa) [inline PWGHF/D2H/Utils/utilsRedDataFormat.h:24]: This should not be defined here, it should rather be an argument of the `checkEvSel` method

**Review by fgrosa — CHANGES_REQUESTED** [review]

File: PWGHF/D2H/Utils/utilsRedDataFormat.h (hunk 5 above)
Comment (Luca610) [reply PWGHF/D2H/Utils/utilsRedDataFormat.h]: Thank you for your comment, I missed it. I fixed it in the current version!

**Review by fgrosa — DISMISSED** [review]

File: PWGHF/D2H/Utils/utilsRedDataFormat.h [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file utilsEvSelHf.h
```
Comment (vkucera) [inline PWGHF/D2H/Utils/utilsRedDataFormat.h:12]:
```suggestion
/// \file utilsRedDataFormat.h
```

**Review by vkucera — CHANGES_REQUESTED** [review]

File: PWGHF/D2H/Utils/utilsRedDataFormat.h [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+#include "PWGHF/Utils/utilsEvSelHf.h"
+
+namespace o2::hf_evsel
+{
+/// Helper function to count collisions at different event selection stages
+/// \tparam useEvSel use information from the EvSel table
+/// \tparam centEstimator centrality estimator
+/// \param collision collision to test against the selection criteria
+template <bool useEvSel, o2::hf_centrality::CentralityEstimator centEstimator, typename Coll>
+void checkEvSel(Coll const& collision, o2::hf_evsel::HfEventSelection& hfEvSel, int& zvtxColl, int& sel8Coll, int& zvtxAndSel8Coll, int& allSelColl)
```
[the remark this answers is not in the document]
Comment (Luca610) [reply PWGHF/D2H/Utils/utilsRedDataFormat.h:29]: I am a compilation error when setting hfEvsel to constant...

[the remark this answers is not in the document]
Comment (vkucera) [reply PWGHF/D2H/Utils/utilsRedDataFormat.h:29]: What error is it?

**Review by vkucera — APPROVED** [review]

**Pull request merged**